server/aesdsocket.c

```diff
@@ -0,0 +1,304 @@
+
+	/* Error occurred in listen, return -1 on error */
+	if(rc == -1)
+	{
+		printf("listen failed\n");
+		return -1;
+	}
+
+	/* Open file for writing */
+	data_file = open(FILE_PATH, O_CREAT | O_APPEND | O_RDWR, S_IRWXU);
+
+	if(data_file == -1)
+	{
+		printf("open failed\n");
+		return -1;
+	}
+
+	/* For Signal Masking during recv and send */
+	sigemptyset(&new_set);
+	sigaddset(&new_set, SIGINT);
+	sigaddset(&new_set, SIGTERM);
+	
+	/* Malloc for recieve buffer from recv and transmit buffer for send */
+	if((rxbuf = (char *)malloc(BUF_SIZE*sizeof(char))) == NULL)
+	{
+		printf("malloc failed");
+		return -1;
+	}	
+
+	if((txbuf = (char *)malloc(BUF_SIZE*sizeof(char))) == NULL)
+	{
+		printf("malloc failed");
+		return -1;
+	}
+	
+	if(startdaemon)
+	{
+		pid = fork();
+		if(pid == -1)
+			return -1;
+
+		else if(pid != 0)
+			exit(0);
+
+		if(setsid() == -1)
+			return -1;
+
+		if(chdir("/") == -1)
+			return -1;
+
+		open("/dev/null", O_RDWR);
+		dup(0);
+		dup(0);
+	}
+	
+	while(1)
+	{
+		/* Get the accepted client fd */
+		client_fd = accept(socket_fd, (struct sockaddr *)&clientsockaddr, &addrsize); 
+
+		/* Error occurred in accept, return -1 on error */
+		if(client_fd == -1)
+		{
+			printf("accept failed\n");
+			free(rxbuf);
+			return -1;
+		}
+
+		/* sockaddr to IP string */
+		char *IP = inet_ntoa(clientsockaddr.sin_addr);
+		/* Log connection IP */
+		syslog(LOG_DEBUG, "Accepted connection from %s\n", IP);
+
+		/* Don't accept a signal while recieving data */
+		if((rc = sigprocmask(SIG_BLOCK, &new_set, &old_set)) == -1)
+			printf("sigprocmask failed\n");
+		
+		/* Initially, write to starting of buffer */
+		bufloc = 0;	
+		
+		/* Receive data */
+		while((recv_bytes = recv(client_fd, rxbuf + bufloc, BUF_SIZE, 0)) > 0)
+		{
+			/* Error occurred in recv, log error */
+			if(recv_bytes == -1)
+			{
+				printf("recv failed, %s\n", strerror(errno));
+				return -1;
+			}
+
+			/* Detect newline character */
+			char* newlineloc = strchr(rxbuf, '\n');
+
+			bufloc = bufloc + recv_bytes;
+
+			/* Buffer size needs to be increased */
+			if((bufloc) >= (bufcount * BUF_SIZE))
+			{
+				bufcount++;
+				char* newptr = realloc(rxbuf, bufcount*BUF_SIZE*sizeof(char));
+
+				if(newptr == NULL)
+				{
+					free(rxbuf);
+					printf("Reallocation failed\n");
+					return -1;
+				}
+
+				else rxbuf = newptr;
+					
+			}
+
+			/* Newline exists, break out here */
+			else if (newlineloc != NULL)
+			       	break;
+			
+		}
+		
+		/* Unmask pending signals */
+		if((rc = sigprocmask(SIG_UNBLOCK, &old_set, NULL)) == -1)
+			printf("sigprocmask failed\n");
+		
+		/* Write to file */
+		int wr_bytes = write(data_file, rxbuf, bufloc);
+
+		if(wr_bytes != bufloc)
+			syslog(LOG_ERR, "Bytes written to file do not match bytes recieved from connection\n");
+	
+		/* Find file size for reallocating write buffer */
+		filesize = lseek(data_file, 0, SEEK_CUR);
+		
+		char* newptr = realloc(txbuf, filesize*sizeof(char));
```

Comment:
@DhruvHMehta what assumption does this make about the amount of available RAM on the heap vs the total amount of packets sent?

Reply:
It assumes that I have enough space on the RAM to allocate all packets (on RAM) and send them all at once. I think I have fixed it the latest commits now by re-writing the send logic to send packet-by-packet and re-use the buffer for each packet. So the maximum size allocated is the size of the biggest packet.